### PR TITLE
fix macros `@which`, `@edit`, `@functionloc`, `@less` for `literal_pow` case.

### LINKED ISSUE #21014: `@which` and `@code_warntype` don't show `^Val{p}` lowering
Cross-posting from discourse: https://discourse.julialang.org/t/incorrect-results-from-which-and-code-warntype-for-literal-powers/2630

I'm trying to use the new literal exponent behavior introduced in #20530 and #20889. I think I'm correctly overloading `literal_pow`, but various macros like `@which`, `@code_warntype` and others all seem to be showing me the wrong behavior. 

```julia
julia> versioninfo()
Julia Version 0.6.0-pre.alpha.136
Commit 6eeabd8 (2017-03-13 00:18 UTC)
Platform Info:
  OS: macOS (x86_64-apple-darwin13.4.0)
  CPU: Intel(R) Core(TM) i7-2860QM CPU @ 2.50GHz
  WORD_SIZE: 64
  BLAS: libopenblas (USE64BITINT DYNAMIC_ARCH NO_AFFINITY Sandybridge)
  LAPACK: libopenblas64_
  LIBM: libopenlibm
  LLVM: libLLVM-3.9.1 (ORCJIT, sandybridge)

julia> immutable MyType
           x::Int
       end

julia> import Base: literal_pow

julia> literal_pow(^, m::MyType, ::Type{Val{p}}) where p = m.x + p
literal_pow (generic function with 6 methods)

julia> m = MyType(1)
MyType(1)

julia> m^2
3
```

That's all good so far. But none of the introspection macros seem to work for `m^2`:

```julia
julia> @which m^2
^(x, p::Integer) in Base at intfuncs.jl:196

julia> @code_warntype m^2
Variables:
  #self#::Base.#^
  x::MyType
  p::Int64

Body:
  begin
      return $(Expr(:invoke, MethodInstance for power_by_squaring(::MyType, ::Int64), :(Base.power_by_squaring), :(x), :(p)))
  end::Any

julia> @code_lowered m^2
CodeInfo(:(begin
        nothing
        return (Base.power_by_squaring)(x, p)
    end))

julia> using Base.Test

julia> @inferred m^2
ERROR: MethodError: no method matching *(::MyType, ::MyType)
Closest candidates are:
  *(::Any, ::Any, ::Any, ::Any...) at operators.jl:424
Stacktrace:
 [1] power_by_squaring(::MyType, ::Int64) at ./intfuncs.jl:166
 [2] ^(::MyType, ::Int64) at ./intfuncs.jl:196
```

are the special lowering rules not being applied inside these macros?

### LINKED ISSUE #43337: `InteractiveUtils` methods show incorrect method for negative powers
```julia
julia> @code_typed true ^ -1
CodeInfo(
1 ─ %1 = Base.slt_int(p, 0)::Bool
└──      goto #5 if not %1
2 ─ %3 = Base.not_int(x)::Bool
└──      goto #4 if not %3
3 ─      invoke Base.throw_domerr_powbysq(_2::Bool, _3::Int64)::Union{}
└──      unreachable
4 ─      nothing::Nothing
5 ┄ %8 = (p === 0)::Bool
│   %9 = Base.or_int(%8, x)::Bool
└──      goto #6
6 ─      return %9
) => Bool

# as expected
julia> a = true ^ 1; println(a, " ", typeof(a))
true Bool

# ???
julia> a = true ^ -1; println(a, " ", typeof(a))
1.0 Float64

# strangely, these forms work
julia> a = true ^ Int64(-1); println(a, " ", typeof(a))
true Bool
julia> a = true ^ -Int64(1); println(a, " ", typeof(a))
true Bool
```

### LINKED ISSUE #53691: `@which 2^(-1)` directs to incorrect method
```julia
julia> 2^(-1)
0.5

julia> @which 2^(-1)
^(x::T, p::T) where T<:Integer
     @ Base intfuncs.jl:310
```
https://github.com/JuliaLang/julia/blob/1ba83f021ecf09a431f7ecd7f35072e6849204eb/base/intfuncs.jl#L348
```julia
julia> Base.power_by_squaring(2,-1)
ERROR: DomainError with -1:
Cannot raise an integer x to a negative power -1.
Make x or -1 a float by adding a zero decimal (e.g., 2.0^-1 or 2^-1.0 instead of 2^-1)or write 1/x^1, float(x)^-1, x^float(-1) or (x//1)^-1.
Stacktrace:
 [1] throw_domerr_powbysq(::Int64, p::Int64)
   @ Base .\intfuncs.jl:265
 [2] power_by_squaring(x_::Int64, p::Int64)
   @ Base .\intfuncs.jl:286
 [3] top-level scope
   @ REPL[79]:1
```

It's because `2^(-1)` is done through `literal_pow`.

`@code_lowered` and `@less` have the same problem.
```julia
julia> @code_lowered 2^(-1)
CodeInfo(
1 ─ %1 = Base.power_by_squaring(x, p)
└──      return %1
)
```

### PR DESCRIPTION
**Pull request by KlausC**
The macros `@which`, `@edit`, `@functionloc`, `@less` from `InteractiveUtils`, if 
applied to the case of literal powers, like `a^12` or `2^-1` used to direct the
user to function `^`, while the compiler generates code for `Base.literal_pow`.

Now the user is shown the code, the compiler generates.

Fixes #53691
Fixes #43337
Fixes #21014

Observation: the `literal_pow` code is generated only if `x` is a signed integer constant with `-2^61 <= x < 2^61`.

### CODE REVIEW

**Comment by KlausC** [comment]
Can somebody have a look at this tiny but valuable improvement of `InteractiveUtils`?

**Comment by mbauman** [comment]
I think this is great and much needed, but I'm quite puzzled why `typemax(Int)÷4` is the cutoff in the first place.  The "real" transformation happens here — or so I think:

https://github.com/JuliaLang/julia/blob/09400e4f5c3e31ba528c2dbcf8e1ccef2a66ce3d/src/julia-syntax.scm#L2683-L2685

So why is `1^2305843009213693951` literal_pow but `1^2305843009213693952` not?  Both are integers, and lisp's `(integer? 2305843009213693952)` agrees.

But flisp does use seemingly-different literal type for values between 2^61-2^63.  And depending on the path we take to lowering, I think these special literals can become something not-integer (specifically within `jl_expand_macros`)?

```
▶ julia --lisp
> 2305843009213693951
2305843009213693951

> 2305843009213693952
#int64(2305843009213693952)

> 9223372036854775807
#int64(9223372036854775807)

> 9223372036854775808
#uint64(9223372036854775808)
```

This all seems very odd.

**Comment by KlausC** [comment]
@mbauman: Yes, this PR is to reflect the current behavior of the compiler(s). Of course that has to be adapted, in case that should change.

**Comment by KlausC** [comment]
# bump

File: stdlib/InteractiveUtils/test/runtests.jl [hunk 1]
```diff
@@ -279,14 +279,49 @@ let x..y = 0
...
+    @test (@which 2^-9223372036854775809).name === :^
+    @test (@which 2^-9223372036854775808).name === :literal_pow
+    @test (@which 2^9223372036854775807).name === :literal_pow
+    @test (@which 2^9223372036854775808).name === :^
+else
+    # literal_pow only for exponents x: -2^31 <= x < 2^31 #53860 (all Int)
+    @test (@which 2^-2147483649).name === :^
+    @test (@which 2^-2147483648).name === :literal_pow
+    @test (@which 2^-2147483647).name === :literal_pow
+    @test (@which 2^-2147483648).name === :^
```
Comment (mbauman) [inline stdlib/InteractiveUtils/test/runtests.jl:315]:
```suggestion
    @test (@which 2^2147483647).name === :literal_pow
    @test (@which 2^2147483648).name === :^
```

This is the test failure in win32.

**Pull request merged**